source/common/common/utility.h

```diff
@@ -636,15 +636,100 @@ class WelfordStandardDeviation {
+  void set(uint8_t branch, std::unique_ptr<FastTrieEntry<Value>> entry) {
+    entries_[branch] = std::move(entry);
+  }
   Value value_{};
-  std::array<std::unique_ptr<TrieEntry>, 256> entries_;
+
+private:
+  std::array<std::unique_ptr<FastTrieEntry>, 256> entries_;
+};
+
+// A SmallTrieEntry aims to be a good balance of performant and
+// space-efficient, by allocating a vector the size of the range of children
+// the node contains. This should be good for most use-cases.
+//
+// For example, a node with children 'a' and 'z' will contain a vector of
+// size 26, containing two values and 24 nulls. A node with only one
+// child will contain a vector of size 1. A node with no children will
+// contain an empty vector.
+//
+// Compared to allocating 256 entries for every node, this makes insertions
+// a little bit inefficient (especially insertions in reverse order), but
+// trie lookups remain O(length-of-longest-matching-prefix) with just a
+// couple of very cheap operations extra per step.
+//
+// By size, having 256 entries for every node makes each node's overhead
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
+template <class Value> class SmallTrieEntry {
+public:
+  Value value_{};
+  SmallTrieEntry* operator[](uint8_t i) const {
+    if (i >= min_child_ && i < min_child_ + children_.size()) {
+      return children_[i - min_child_].get();
+    }
+    return nullptr;
+  }
+
+  void set(uint8_t branch, std::unique_ptr<SmallTrieEntry<Value>> entry) {
+    if (children_.empty()) {
+      children_.reserve(1);
+      children_.emplace_back(std::move(entry));
+      min_child_ = branch;
+      return;
+    }
+    if (branch < min_child_) {
+      // expand the vector backwards, by moving the existing vector onto
+      // the end of a newly allocated one.
+      std::vector<std::unique_ptr<SmallTrieEntry>> new_children;
+      new_children.reserve(min_child_ - branch + children_.size());
+      new_children.resize(min_child_ - branch);
+      std::move(children_.begin(), children_.end(), std::back_inserter(new_children));
+      new_children[0] = std::move(entry);
+      min_child_ = branch;
+      children_ = std::move(new_children);
+      return;
+    }
+    if (branch >= min_child_ + children_.size()) {
+      // expand the vector forwards
+      children_.resize(branch - min_child_ + 1);
+      // fall through to "insert" behavior.
+    }
+    children_[branch - min_child_] = std::move(entry);
+  }
+
+private:
+  uint8_t min_child_{0};
+  std::vector<std::unique_ptr<SmallTrieEntry>> children_;
 };
 
 /**
  * A trie used for faster lookup with lookup time at most equal to the size of the key.
  */
-template <class Value> struct TrieLookupTable {
+template <template <class> class EntryType, class Value> struct TrieLookupTable {
```

Comment:
If you do it that way then at the use-sites you have to declare like

`TrieLookupTable<WhateverTrieEntry<std::unique_ptr<X>>, std::unique_ptr<X>>`

versus this way it's

`TrieLookupTable<WhateverTrieEntry, std::unique_ptr<X>>`

Reply:
What do you think of this idiom:
```
template<class ValueType> class DenseEntry {
 public:
  using Value = ValueType;
};
template<class ValueType> class SparseEntry {
 public:
  using Value = ValueType;
};

template<class EntryType> class MyContainer {
  using Value = typename EntryType::Value;
};

MyContainer<SparseEntry<int>> sparse;
MyContainer<DenseEntry<bool>> dense;
```

I'm fine with whawt you have; just wanted to put this out there as a taste test. I'm not sure why I can't directly use SparseEntry::ValueType from MyContainer without the 'using' nickname.

Reply:
Just to give another alternative, I slightly prefer using explicit aliases instead of letting the user of TrieLookupTable know the possible types of the first parameter.
What I mean is defining the following:
```
template <typename T>
using BigTrieLookupTable = TrieLookupTable<BigTrieEntry, T>;

template <typename T>
using SmallTrieLookupTable = TrieLookupTable<SmallTrieEntry, T>;
```